SofaKernel/modules/SofaCore/src/sofa/core/objectmodel/DDGLink.h

```diff
@@ -0,0 +1,126 @@
+#pragma once
+
+#include <sofa/core/objectmodel/DDGNode.h>
+#include <sofa/core/objectmodel/Base.h>
+#include <sofa/core/ExecParams.h>
+
+namespace sofa
+{
+namespace core
+{
+namespace objectmodel
+{
+
+
+/**
+ * @brief The BaseLink class
+ *
+ * BaseLink inherits DDGNode, thus is part of the data dependency graph, thus can have inputs and outputs.
+ * When setting a link, the linked base's componentState data is added as an input to the BaseLink,
+ * which creates the connection between the BaseLink and the DDG.
+ * any data, engine, etc. can then be connected as output.
+ */
+class BaseDDGLink : public DDGNode
+{
+public:
+    /// Flags that describe some properties of a Data, and that can be OR'd together.
+    /// \todo Probably remove FLAG_PERSISTENT, FLAG_ANIMATION_INSTANCE, FLAG_VISUAL_INSTANCE and FLAG_HAPTICS_INSTANCE, it looks like they are not used anywhere.
+    enum DataFlagsEnum
+    {
+        FLAG_NONE       = 0,      ///< Means "no flag" when a value is required.
+        FLAG_READONLY   = 1 << 0, ///< The Data will be read-only in GUIs.
+        FLAG_DISPLAYED  = 1 << 1, ///< The Data will be displayed in GUIs.
+        FLAG_PERSISTENT = 1 << 2, ///< The Data contains persistent information.
+        FLAG_AUTOLINK   = 1 << 3, ///< The Data should be autolinked when using the src="..." syntax.
+        FLAG_REQUIRED = 1 << 4, ///< True if the Data has to be set for the owner component to be valid (a warning is displayed at init otherwise)
+        FLAG_ANIMATION_INSTANCE = 1 << 10,
+        FLAG_VISUAL_INSTANCE = 1 << 11,
+        FLAG_HAPTICS_INSTANCE = 1 << 12,
+    };
+    /// Bit field that holds flags value.
+    typedef unsigned DataFlags;
+
+    /// Default value used for flags.
+    enum { FLAG_DEFAULT = FLAG_DISPLAYED | FLAG_PERSISTENT | FLAG_AUTOLINK };
+
+    /// This internal class is used by the initLink() methods to store initialization parameters of a Data
+    class InitDDGLink
+    {
+    public:
+        InitDDGLink()
+            : name(""),
+              help(""),
+              group(""),
+              linkedBase(nullptr),
+              owner(nullptr),
+              dataFlags(FLAG_DEFAULT) {}
+        std::string name;
+        std::string help;
+        std::string group;
+        Base* linkedBase;
+        Base* owner;
+        DataFlags dataFlags;
+    };
+
+    explicit BaseDDGLink(const InitDDGLink& init);
+
+    virtual ~BaseDDGLink() override;
+
+    void setOwner(Base* owner);
+
+    void set(Base* linkedBase);
+
+    Base* get();
```

Comment:
I would add a `const Base * get() const;` in case I create a const link `DDGLink<const MyComponent>`

Reply:
Done. I'm not super happy about the design though: The setter taking a const Base* stores the const_cast'ed pointer internally and prays nobody calls the non-const getter afterwards.

When manipulating DDGLink<T> objects, that's not an issue because the wrong use case is encapsulated in the child class that guarantees it doesn't happen. But nothing prevents someone from manipulating a DDTLink<const T> through its abstract type BaseDDGLink and calling the non-const getter, which would lead to undefined behaviors.

@jnbrunet I'd like a second pair of eyes on that one, if you have some time, in case you would see a way to fix that..?

Reply:
Hey @marques-bruno 

I think the idea of having a "const DDGLink" is not really that the DDGLink is declared as `const`, but more that its linked `T` object is. Since the `DDGLink::get` function cannot realistically be marked `const` (because it updates its data counters), you would not have a `const DDGLink<T>` but instead have a `DDGLink<const T>`.

It think this is how I would do it:
```
// The BaseDDGLink does not actually contains a linked object
class BaseDDGLink : public DDGNode
{
public:
// (...)
    void set(Base* linkedBase) = 0; // Pure virtual, must be override by DDGLink<T>
    // !! Remove the const  version of set from here

    const Base* get() = 0; // Pure virtual, must be override by DDGLink<T>
    // !! No non-const getter here !!
    
// (...)
}

// Non-Const version
template <class T>
class DDGLink : public BaseDDGLink
{
  public:
// (...)
    void set(T* linkedBase) final;
    T * get(); // Here the owned of the DDGLink will be able to modify the linked object T
    const T * get() final;
// (...)

private:
    T * m_linkedBase {nullptr};
}

// Const version
template <class T>
class DDGLink<const T> : public BaseDDGLink
{
  public:
// (...)
    void set(T* linkedBase) final;
    void set(const T* linkedBase);
    const T * get() final;
// (...)
private:
    const T * m_linkedBase {nullptr};
}
```

What do you think?

Reply:
Hey! So I'm gonna revert the latest changes (the handling of const template types in DDGLinks), because clearly the implementation really sucks and adds more problems than it solves...

@jnbrunet I personally don't have a use case for DDGLink<const T*>, and I'm wondering if enforcing constness here isn't a bit of a mistake. Sure it would enforce and explicit the fact that a component cannot modify another one through its link, but does this issue really occurred already?
I think the same applies for datafields. We *could* imagine a use case where a datafield should be readonly, but maybe we don't need to enforce constness for that. I'm thinking using flags, like we do on data (required, readonly, etc)

Reply:
Well, side effects was pretty much one of the starting point of the discussion on data, no?

Maybe I did not understand correctly our discussions, but I though that we talked about enforcing links to be "const" by "default". This will prevent calling a method of the linked component that would changes its internal state without knowing it and instead use the data mechanism which is made for this.

If your link is "const", then calling a method of the linked object that changes its internal state will produce a compilation error, forcing the developer to think if it is the right way of doing this and also help the PR reviewers (eg: why are you changing the link to a non-const?). 

This should also point out some "limits" of the data mechanism (links that will have to be non-const because datas cannot do its job for this case).